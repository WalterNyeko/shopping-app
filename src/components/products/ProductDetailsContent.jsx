import React, { Fragment } from "react";
import StarRatingComponent from "react-star-rating-component";
import StarRatings from "react-star-ratings";
import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";
import InputLabel from "@material-ui/core/InputLabel";
import MenuItem from "@material-ui/core/MenuItem";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";

const ProductDetailsContent = ({
  product,
  onStarClick,
  rating,
  productReviews,
  sizeOpen,
  colorOpen,
  handleChange,
  handleSubmit,
  handleCloseColor,
  handleOpenColor,
  handleCloseSize,
  handleOpenSize,
  color,
  size,
  colorsData,
  sizesData
}) => {
  const useStyles = makeStyles(theme => ({
    root: {
      padding: theme.spacing(3, 2)
    },
    button: {
      display: "block",
      marginTop: theme.spacing(2)
    },
    formControl: {
      margin: theme.spacing(1),
      minWidth: 120
    }
  }));
  const classes = useStyles();
  return (
    <Fragment>
      {product && product.thumbnail && (
        <div class="container-fluid product-body">
          <div class="row">
            <div class="col-md-5 col-sm-12 text-center">
              <img
                src={require(`../../images/${product.thumbnail}`)}
                alt=""
                width="180px"
              />
            </div>
            <div class="col-md-7">
              <h3>{product.name}</h3>
              <h5 class="text-danger">
                <strike>${product.price}</strike>
              </h5>
              <h3 class="text-danger ml-5">${product.discounted_price}</h3>
              <p>{product.description}</p>
              <h5>Color</h5>
              <FormControl className={classes.formControl}>
                <InputLabel htmlFor="demo-controlled-open-select">
                  Select Color
                </InputLabel>
                <Select
                  open={colorOpen}
                  value={color}
                  onChange={handleChange}
                  onClose={handleCloseColor}
                  onOpen={handleOpenColor}
                  inputProps={{
                    name: "color",
                    id: "demo-controlled-open-select"
                  }}
                >
                  {colorsData &&
                    colorsData.map(
                      ({ attribute_value_id, attribute_value }) => (
                        <MenuItem
                          value={attribute_value}
                          key={attribute_value_id}
                        >
                          {attribute_value}
                        </MenuItem>
                      )
                    )}
                </Select>
              </FormControl>
              <h5>Size</h5>
              <FormControl className={classes.formControl}>
                <InputLabel htmlFor="demo-controlled-open-select">
                  Select Size
                </InputLabel>
                <Select
                  open={sizeOpen}
                  value={size}
                  onChange={handleChange}
                  onClose={handleCloseSize}
                  onOpen={handleOpenSize}
                  inputProps={{
                    name: "size",
                    id: "demo-controlled-open-select1"
                  }}
                >
                  {sizesData &&
                    sizesData.map(({ attribute_value_id, attribute_value }) => (
                      <MenuItem
                        value={attribute_value}
                        key={attribute_value_id}
                      >
                        {attribute_value}
                      </MenuItem>
                    ))}
                </Select>
              </FormControl>
              <button
                className="btn btn-danger text-center ml-5 mt-3"
                onClick={handleSubmit}
              >
                Add To Cart
              </button>
            </div>
          </div>
          <h3 class="text-center mt-5">Leave a Review</h3>
          <div class="row mb-4">
            <div class="col-md-8 mx-auto">
              <form className="mb-2">
                <textarea type="text" class="form-control" />
              </form>
              <div className="text-center">
                <div className="m-2">
                  <StarRatings
                    rating={rating}
                    starRatedColor="green"
                    changeRating={onStarClick}
                    numberOfStars={5}
                    name="rating"
                  />
                </div>
                <br />
                <button className="btn btn-danger text-center">
                  Leave a Review
                </button>
              </div>
            </div>
          </div>

          <div className="row">
            <div className="col-md-3" />
            <div className="col-md-6">
              {productReviews &&
                productReviews.map(({ name, review, rating, created_on }) => (
                  <Paper className={`${classes.root} m-4`}>
                    <Typography variant="h5" component="h3">
                      {name}
                    </Typography>
                    <Typography component="p">{review}</Typography>
                    <StarRatingComponent
                      name="rate1"
                      starCount={5}
                      value={rating}
                      starColor="green"
                      className="mt-2"
                    />
                    <Typography component="p">
                      Reviewed on {created_on}
                    </Typography>
                  </Paper>
                ))}
            </div>
            <div className="col-md-3" />
          </div>
        </div>
      )}
    </Fragment>
  );
};
export default ProductDetailsContent;
