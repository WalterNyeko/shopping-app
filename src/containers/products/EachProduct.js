import React, { Component, Fragment } from "react";
import EachProductComponent from "../../components/products/EachProduct";
import Pagination from "material-ui-flat-pagination";
import { connect } from "react-redux";
import {
  getProducts,
  getProductsPerDepartment,
  getProductsPerCategory,
  getProductsAfterSearch
} from "../../store/actions/Products";

export class EachProduct extends Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: false,
      offset: 0,
      limit: 8
    };
  }

  handleClick = offset => {
    const { limit } = this.state;
    const currentPage = offset / limit + 1;
    this.setState({ offset });
    const { getProducts } = this.props;
    const requestData = {
      page: currentPage,
      limit,
      description_length: 100
    };
    getProducts(requestData);
    if (this.props.departmentId) {
      const { departmentId, getProductsPerDepartment } = this.props;
      const requestDataDepartment = {
        page: currentPage,
        limit,
        description_length: 100,
        departmentId
      };
      getProductsPerDepartment(requestDataDepartment);
    }

    if (this.props.categoryId) {
      const { categoryId, getProductsPerCategory } = this.props;
      const requestDataCategory = {
        page: currentPage,
        limit,
        description_length: 100,
        categoryId
      };
      getProductsPerCategory(requestDataCategory);
    }

    if (this.props.afterSearch) {
      const { afterSearch, getProductsAfterSearch } = this.props;
      const requestDataSearch = {
        page: currentPage,
        limit,
        description_length: 100,
        query_string: afterSearch
      };
      getProductsAfterSearch(requestDataSearch);
    }
  };

  render() {
    const { products } = this.props;
    return (
      <Fragment>
        <div className="w-100 text-center mb-3">
          <div className="col-md-12 mx-auto">
            <Pagination
              limit={this.state.limit}
              offset={this.state.offset}
              total={products && products.count ? products.count : 1}
              onClick={(event, offset) => this.handleClick(offset)}
            />
          </div>
        </div>
        <div className="row">
          <EachProductComponent products={products} />
        </div>
      </Fragment>
    );
  }
}
export default connect(
  null,
  {
    getProducts,
    getProductsPerDepartment,
    getProductsPerCategory,
    getProductsAfterSearch
  }
)(EachProduct);
