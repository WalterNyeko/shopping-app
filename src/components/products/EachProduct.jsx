import React, { Fragment } from 'react';
import { truncateWords } from '../../helpers/index';
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import { useStyles } from '../../styles/Products';
import Grid from '@material-ui/core/Grid';
import { Modal } from 'antd';


const EachProduct = ({ 
    products,
    modalTitle, 
    modalContent, 
    modalWidth,
    handleOk, 
    handleCancel,
    showModal,
    visible, 
    }) => {
    const classes = useStyles();

    return (
        <Fragment>
            {products && products.rows? products.rows.map(({product_id, name, description, price, discounted_price }) => (
              <Fragment>
              <Grid item key={product_id} xs={6} sm={3} md={3} onClick={showModal}>
                <Card className={classes.card}>
                    <Typography className={classes.heading} gutterBottom variant="h6" component="h2">
                      {name}
                    </Typography>
                  <CardMedia
                    className={classes.cardMedia}
                    image="https://source.unsplash.com/random"
                    title="Image Title"
                  />
                  <CardContent className={classes.cardContent}>
                    <div className="row">
                      <div className="col-md-6">
                        <Typography className="text-danger">
                          <strike>$ {price}</strike>
                        </Typography>
                      </div>
                      <div className="col-md-6">
                        <Typography className="btn btn-danger btn-sm" style={{float: 'right'}}>
                          $ {discounted_price}
                        </Typography>
                      </div>
                    </div>
                    <br/>
                    
                    <Typography>
                      {truncateWords(description, 20)}...
                    </Typography>
                  </CardContent>
                </Card>
              </Grid>
              </Fragment>
            )): ('')}
            <Modal
                 title={modalTitle}
                 visible={visible}
                 onOk={handleOk}
                 onCancel={handleCancel}
                 width={modalWidth}
                 footer={null}
              >
                 <p>{modalContent}</p>
              </Modal>
        </Fragment>
    )
}
export default EachProduct;