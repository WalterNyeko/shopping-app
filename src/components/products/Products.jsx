import React from "react";
import CssBaseline from "@material-ui/core/CssBaseline";
import Grid from "@material-ui/core/Grid";
import Container from "@material-ui/core/Container";
import { useStyles } from "../../styles/Products";
import LinearProgress from "@material-ui/core/LinearProgress";
import EachProduct from "../../containers/products/EachProduct";

const Products = ({ products, departmentId, categoryId, afterSearch }) => {
  const classes = useStyles();
  return (
    <React.Fragment>
      <CssBaseline />
      <main>
        <Container className={classes.cardGrid}>
          <Grid container spacing={3}>
            <EachProduct
              products={products}
              modalTitle="Test"
              modalContent=""
              modalWidth="600px"
              visible
              departmentId={departmentId}
              categoryId={categoryId}
              afterSearch={afterSearch}
            />
          </Grid>
        </Container>
      </main>
      {products && products.rows ? (
        ""
      ) : (
        <LinearProgress
          classes={{
            colorPrimary: classes.linearColorPrimary,
            barColorPrimary: classes.linearBarColorPrimary
          }}
        />
      )}
    </React.Fragment>
  );
};
export default Products;
