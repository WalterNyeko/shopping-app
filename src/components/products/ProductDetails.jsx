import React from "react";
import CssBaseline from "@material-ui/core/CssBaseline";
import Container from "@material-ui/core/Container";
import { useStyles } from "../../styles/Products";
import LinearProgress from "@material-ui/core/LinearProgress";

const ProductDetails = ({ product }) => {
  const classes = useStyles();

  return (
    <React.Fragment>
      <CssBaseline />
      <main>
        <Container className={classes.cardGrid} />
      </main>
      {product ? (
        ""
      ) : (
        <LinearProgress
          classes={{
            colorPrimary: classes.linearColorPrimary,
            barColorPrimary: classes.linearBarColorPrimary
          }}
        />
      )}
    </React.Fragment>
  );
};
export default ProductDetails;
