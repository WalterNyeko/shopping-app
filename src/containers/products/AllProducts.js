import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import ProductsComponent from "../../components/products/Products";
import { getProducts } from "../../store/actions/Products";

export class AllProducts extends Component {
  componentWillMount = () => {
    const { getProducts } = this.props;
    const requestData = {
      page: 1,
      limit: 8,
      description_length: 100
    };
    getProducts(requestData);
  };

  render() {
    let theProducts;
    const { products, productsAfterSearch } = this.props.allProducts;
    if (productsAfterSearch && productsAfterSearch.rows) {
      theProducts = productsAfterSearch;
    } else {
      theProducts = products;
    }
    return (
      <Fragment>
        <ProductsComponent
          products={theProducts}
          afterSearch={this.props.afterSearch}
        />
      </Fragment>
    );
  }
}
const mapStateToProps = state => ({
  allProducts: state.products
});
export default connect(
  mapStateToProps,
  { getProducts }
)(AllProducts);
