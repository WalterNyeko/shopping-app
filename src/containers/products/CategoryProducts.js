import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import ProductsComponent from '../../components/products/Products';
import { getProductsPerCategory } from '../../store/actions/Products';

class CategoryProducts extends Component {

    componentWillMount = () => {
        const { getProductsPerCategory, categoryId } = this.props;
        getProductsPerCategory(categoryId);
    } 

    componentWillReceiveProps = (nextProps) => {
        if(nextProps.categoryId !== this.props.categoryId){
            const { getProductsPerCategory, categoryId } = nextProps;
            getProductsPerCategory(categoryId);
        }
 
    }
    
    render() {
        const { productsPerCategory } = this.props.allProducts;
        return (
            <Fragment>
                <ProductsComponent 
                    products={productsPerCategory}/>
            </Fragment>
        )
    }
}
const mapStateToProps = (state) => ({
    allProducts: state.products
})
export default connect(mapStateToProps, { getProductsPerCategory })(CategoryProducts);