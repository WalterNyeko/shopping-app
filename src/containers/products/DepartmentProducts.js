import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import ProductsComponent from "../../components/products/Products";
import { getProductsPerDepartment } from "../../store/actions/Products";

export class DepartmentProducts extends Component {
  componentWillMount = () => {
    const { getProductsPerDepartment, departmentId } = this.props;
    const requestData = {
      page: 1,
      limit: 8,
      description_length: 100,
      departmentId
    };
    getProductsPerDepartment(requestData);
  };

  componentWillReceiveProps = nextProps => {
    if (nextProps.departmentId !== this.props.departmentId) {
      const { getProductsPerDepartment, departmentId } = nextProps;
      const requestData = {
        page: 1,
        limit: 8,
        description_length: 100,
        departmentId
      };
      getProductsPerDepartment(requestData);
    }
  };

  render() {
    const { productsPerDepartment } = this.props.allProducts;
    return (
      <Fragment>
        <ProductsComponent
          products={productsPerDepartment}
          departmentId={this.props.departmentId}
          afterSearch={this.props.afterSearch}
        />
      </Fragment>
    );
  }
}
const mapStateToProps = state => ({
  allProducts: state.products
});
export default connect(
  mapStateToProps,
  { getProductsPerDepartment }
)(DepartmentProducts);
