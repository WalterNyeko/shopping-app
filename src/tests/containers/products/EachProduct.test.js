import React from "react";
import { shallow, mount } from "enzyme";
import { Router } from "react-router";
import { EachProduct } from "../../../containers/products/EachProduct";
import EachProductComponent from "../../../components/products/EachProduct";
import history from "../../../helpers/history";

describe("Each Product", () => {
  const props = {
    products: {
      count: 10,
      rows: [{ product_id: 1, thumbnail: "altar-piece.gif" }]
    }
  };
  const wrapper = shallow(<EachProduct {...props} />);
  it("should render without crashing", () => {
    expect(wrapper).toMatchSnapshot();
  });

  const componentProps = {
    products: {
      count: 11,
      rows: [
        {
          product_id: 1,
          thumbnail: "altar-piece.gif",
          description: "Test one two"
        }
      ]
    },
    getProducts: jest.fn(),
    getProductsPerDepartment: jest.fn(),
    departmentId: 1,
    getProductsPerCategory: jest.fn(),
    categoryId: 1,
    getProductsAfterSearch: jest.fn(),
    afterSearch: true
  };
  const componentWrapper = mount(
    <Router history={history}>
      <EachProductComponent {...componentProps} />
    </Router>
  );
  it("should render component without crashing", () => {
    const wrapper = shallow(<EachProduct {...componentProps} />);
    const spy = jest.spyOn(wrapper.instance(), "handleClick");
    wrapper.instance().handleClick(10);
    expect(spy).toHaveBeenCalled();
  });
});
