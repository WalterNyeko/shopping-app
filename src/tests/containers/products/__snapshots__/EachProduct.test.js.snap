// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Each Product should render without crashing 1`] = `ShallowWrapper {}`;
